torch/_inductor/kernel/flex_attention.py

```diff
@@ -22,9 +35,85 @@ def sdpa_grid(batch_size, num_heads, num_queries, d_model, meta):
     return (triton.cdiv(num_queries, meta["BLOCK_M"]), batch_size * num_heads, 1)
 
 
-sdpa_template = TritonTemplate(
-    name="sdpa",
-    grid=sdpa_grid,
+def create_placeholder(
+    name: str, dtype: torch.dtype, device: torch.device
+) -> TensorBox:
+    """Creates a placeholder input buffers for producing subgraph_output."""
+    input_buffer = InputBuffer(name, FixedLayout(device, dtype, [1], [1]))
+    return TensorBox.create(input_buffer)
+
+
+def build_subgraph_buffer(
+    args: Tuple[IRNode],
+    placeholder_inps: List[TensorBox],
+    subgraph: Subgraph,
+    is_joint_graph: bool,
+) -> ComputedBuffer:
+    """This function's goal is to take in the required args and produce the subgraph buffer
+    The subgraph buffer is a ComputedBuffer that will be inlined into the triton template
+
+    Args:
+        args: The args that were passed into the flex_attention kernel
+        subgraph: The Subgraph ir for which to produce the output node
+        placeholder_inps: The list of scalar inputs, these were created on the fly through `create_placeholder`
+        is_joint_graph: Whether or not this subgraph represents the joint graph
+    """
+    cnt = 0
+    env = {}
+    for node in subgraph.graph_module.graph.nodes:
+        # There are two classes of placeholder inpts that we need
+        # to handle differently. For the first n_scalar_inps inputs
+        # we expect that these placeholders were generated by the make_fx call
+        # in the flex Attention HOP. So we need to create a new placeholder
+        # TensorBox for each of these inputs. For the rest of the inputs we
+        # expect that these are lifted inputs that fill up the '*other_buffers'
+        # tuple and already have corresponding TensorBoxes passed in as args.
+        if node.op == "placeholder":
+            is_lifted_input = cnt >= len(placeholder_inps)
+            env[node] = args[cnt - 1] if is_lifted_input else placeholder_inps[cnt]
+            cnt += 1
+        elif node.op == "call_function":
+            # For call_function we use the default lowerings and pass in the
+            # already created TensorBoxes as args
+            from torch.utils._pytree import tree_map
+
+            env[node] = lowerings[node.target](
+                *tree_map(lambda x: env[x] if x in env else x, node.args)
+            )
+        elif node.op == "output":
+            # For the output node we need to create a ComputedBuffer
+            # which represents the actual score modification
+            # TODO confirm why For the joint graph we expect the output nodes args to be
```

Comment:
This is because the forward of score mod takes 5 arguments, and only the first is differentiable. Thus the first arg gets its grad while the rest get Nones